[CON-961] Update the docsite to include how Mail works with the KDS

ac101m opened the pull request:
# Changes
- Fixed outdated information in the docsite regarding the KDS.
- Added section for encrypting mail keys.

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -54,147 +46,102 @@ enclaveHost.start(null, null, null, kdsConfiguration) {}
+Just like Conclave attestation, it follows the [enclave constraints DSL](constraints.md).
+
+### Setting the `useOwnCodeHash` and `useOwnCodeSignerAndProductID` properties
+In addition to the `constraint` property, the `policyConstraint` section has two properties which allow you to include the code hash or code signer and product ID of the current enclave to the constraint.
```
The `useOwnCodeHash` and `useOwnCodeSignerAndProductID` properties of the `policyConstraint` section allow you to include the code hash or code signer and product ID of the current enclave to the constraint.

ac101m replied:
I have changed the wording here, but not in exactly the way requested.

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -54,147 +46,102 @@ enclaveHost.start(null, null, null, kdsConfiguration) {}
+
+### Example `PolicyConstraint` configurations
+#### Exact version of an enclave running on a secure platform
+This is an example of a restrictive policy constraint, whereby access to the derived key is restricted to release mode enclaves with a specific code hash running on a fully patched and up-to-date SGX capable host.
```
In this restrictive policy constraint example, access to the derived key is restricted to release mode enclaves with a specific code hash running on a fully patched and up-to-date SGX-capable host.

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -54,147 +46,102 @@ enclaveHost.start(null, null, null, kdsConfiguration) {}
-to deploy the patch to the system, bringing it back to the `SECURE` state.
+This is a restrictive policy constraint with significant operational implications.
+Because of the code hash requirement, persisted data will be inaccessible to subsequent versions of the enclave.
+Finally, if the system is not fully patched, the derived key will not be accessible and the enclave may be denied access to storage keys until patches are installed.
```
Also, if the system...

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -54,147 +46,102 @@ enclaveHost.start(null, null, null, kdsConfiguration) {}
+Finally, if the system is not fully patched, the derived key will not be accessible and the enclave may be denied access to storage keys until patches are installed.
+
+#### Exact version of an enclave running on a platform that may need updating
+This is an example of a slightly less restrictive policy constraint, whereby access to the derived key is permitted in cases where non-critical security patches are available for the host system.
```
In this slightly less restrictive policy example, access to the derived key is permitted in cases where non-critical security patches are available for the host system.

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +154,95 @@ kds {
-generated per product, but shared with all versions of enclaves within a single
-product.
+This avoids the denial of service that happens in the more restrictive case above by giving the enclave host time to patch their systems to bring them back to the `SECURE` state.
+The operational implications of enclave updates are not affected however, as access to the derived key still requires that the enclave have a specific code hash.
```
This constraint doesn't affect the operational implications of enclave updates because access to the derived key still requires that the enclave has a specific code hash.

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +154,95 @@ kds {
-when using an external auditor to validate enclave code and sign the enclave on
-the developer's behalf.
+#### Any enclave for a particular product signed using the same key
+This is an example of a permissive policy constraint, whereby access to the derived key is permitted for enclaves with a matching product ID and signing key.
```
In this permissive policy constraint example, access to the derived key is permitted for enclaves with a matching product ID and signing key.

JohnJoser3 commented on docs/docs/kds-configuration.md:
````diff
@@ -207,32 +154,95 @@ kds {
 ```
+
+This policy makes it easy to migrate encrypted data to a new version of the enclave, as access to the key is permitted to any enclave with a matching signing key and [product ID](enclave-configuration.md#productid).
+This constraint has fewer operational implications, but security is weakened as the enclave signer must be trusted not to sign an unsafe or malicious enclave.
````
This constraint has fewer operational implications. However, this constraint reduces the security, as the enclave signer must be trusted not to sign an unsafe or malicious enclave.

ac101m replied:
Changes made, though not exactly as requested, PTAL

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +154,95 @@ kds {
+This policy makes it easy to migrate encrypted data to a new version of the enclave, as access to the key is permitted to any enclave with a matching signing key and [product ID](enclave-configuration.md#productid).
+This constraint has fewer operational implications, but security is weakened as the enclave signer must be trusted not to sign an unsafe or malicious enclave.
+
+!!!note
```
!!!Note

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +154,95 @@ kds {
+This constraint has fewer operational implications, but security is weakened as the enclave signer must be trusted not to sign an unsafe or malicious enclave.
+
+!!!note
+In this example, security may be enhanced by having an external auditor build and sign the enclave on the behalf of the developer.
```
In this example, you can enhance security by assigning an external auditor to build and sign the enclave on behalf of the developer.

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +154,95 @@ kds {
+!!!note
+In this example, security may be enhanced by having an external auditor build and sign the enclave on the behalf of the developer.
+
+## KDS keys for mail
```
KDS keys for Conclave Mail

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +154,95 @@ kds {
+In this example, security may be enhanced by having an external auditor build and sign the enclave on the behalf of the developer.
+
+## KDS keys for mail
+The KDS is also capable of providing keys for use with [Conclave mail](mail.md) through use of a specially constructed [`PostOffice`](api/-conclave/com.r3.conclave.mail/-post-office/index.html).
```
Conclave Mail

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +154,95 @@ kds {
+The KDS is also capable of providing keys for use with [Conclave mail](mail.md) through use of a specially constructed [`PostOffice`](api/-conclave/com.r3.conclave.mail/-post-office/index.html).
+
+When using a KDS post office, the post office requests a public key from the KDS for use in a Conclave [`PostOffice`](api/-conclave/com.r3.conclave.mail/-post-office/index.html) and specifies which enclaves should be permitted access to the corresponding private key.
+Upon receiving mail, the enclave requests the private key from the same KDS and uses it to decrypt the mail object from the client.
```
Upon receiving Mail
decrypt the Mail object

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +154,95 @@ kds {
+
+When using a KDS post office, the post office requests a public key from the KDS for use in a Conclave [`PostOffice`](api/-conclave/com.r3.conclave.mail/-post-office/index.html) and specifies which enclaves should be permitted access to the corresponding private key.
+Upon receiving mail, the enclave requests the private key from the same KDS and uses it to decrypt the mail object from the client.
+In this manner, users can create mail objects which can be decrypted by any enclave that matches a given set of constraints, without first having to receive and attest to the [`EnclaveInstanceInfo`](api/-conclave/com.r3.conclave.common/-enclave-instance-info/index.html) of a specific enclave.
```
Mail objects

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +154,95 @@ kds {
+In this manner, users can create mail objects which can be decrypted by any enclave that matches a given set of constraints, without first having to receive and attest to the [`EnclaveInstanceInfo`](api/-conclave/com.r3.conclave.common/-enclave-instance-info/index.html) of a specific enclave.
+
+### Configuring the Enclave
+To make use of KDS keys for mail, the `kdsEnclaveConstraints` field in the enclaves `build.gradle` must be configured.
```
Mail

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +154,95 @@ kds {
+
+### Configuring the Enclave
+To make use of KDS keys for mail, the `kdsEnclaveConstraints` field in the enclaves `build.gradle` must be configured.
+This configuration alone is sufficient to use KDS keys with mail.
```
Mail

JohnJoser3 commented on docs/docs/kds-configuration.md:
````diff
@@ -207,32 +154,95 @@ kds {
+...
+```
+
+!!!note
````
!!!Note

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +154,95 @@ kds {
+
+
+### Defining a KDS key spec
+To make use of KDS keys with mail, the client must construct an appropriately configured [`PostOffice`](api/-conclave/com.r3.conclave.mail/-post-office/index.html) object with which to encrypt mail.
```
Capitalize Mail in both places.

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +154,95 @@ kds {
+- `<kds-enclave-constraints>` are the constraints that will be used when attesting to the KDS enclave.
+  These should match the expected details of the KDS.
+
+The [`PostOffice`](api/-conclave/com.r3.conclave.mail/-post-office/index.html) may then be used to encrypt and decrypt messages as usual.
```
You can then use the [`PostOffice`](api/-conclave/com.r3.conclave.mail/-post-office/index.html) to encrypt and decrypt messages as usual.

JohnJoser3 commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +154,95 @@ kds {
+- `policyConstraint` - A string which specifies the constraints that the enclave must meet in order to derive the corresponding private key and decrypt messages. The string should be in the [enclave constraints DSL](constraints.md) format.
+
+### Using the KDS post office builder
+Once an appropriate key spec has been created, a post office can be instantiated using the [`PostOfficeBuilder`](api/-conclave/com.r3.conclave.client/-post-office-builder/index.html) class.
```
After creating an appropriate key spec, you can instantiate a post office using the [`PostOfficeBuilder`](api/-conclave/com.r3.conclave.client/-post-office-builder/index.html) class.

JohnJoser3 reviewed (changes requested):
I like the detailed and step-by-step instructions. Most of my comments are to match the content according to our style guide.

ac101m commented on docs/docs/kds-configuration.md:
```diff
@@ -54,147 +46,102 @@ enclaveHost.start(null, null, null, kdsConfiguration) {}
-[`EnclaveInstanceInfo`](api/-conclave/com.r3.conclave.common/-enclave-instance-info/index.html) remote attestation report for the enclave. This ensures
-the key specification is not tampered with.
+## KDS keys for persistent storage
+To use KDS derived keys for persistent storage, a specification for how storage keys will be derived must also be provided.
```
Changed to "you need to" rather than "you should", as this isn't optional. Reply if you disagree, otherwise resolve this comment :)

shamsasari commented on build.gradle:
```diff
@@ -178,10 +178,10 @@ List<PublishedProject> publishedProjects = [
     dokkaSourceSets {
         register("conclave") {
-            displayName.set("Conclave Core")
+            displayName.set("Conclave")
```
Can you try reverting this back to Conclave Core. I want the title in the API docs to say that.

ac101m replied:
Reverted.
Will raise another PR to fix the broken links.

shamsasari replied:
No, just revert the display name. The other `moduleName` _probably_ refers to the link so change that to Conclave.

ac101m replied:
Doesn't work. The moduleName controls both the title and the links!

shamsasari commented on docs/docs/kds-configuration.md:
```diff
@@ -1,47 +1,39 @@
-## Configuring the KDS URL
-For an enclave to obtain a key from the KDS, the URL of the KDS and the
-specification of the key to obtain must both be configured by the developer.
+!!!Note
+The URL for the R3 public KDS is `https://kds.dev.conclave.cloud`.
```
This is not the correct syntax. You can see how it's rendered by looking at the docsite output from TC: https://r3-conclave.artifacts.teamcity.com/presignedTokenAuth/fbffc162-7a85-43ef-8c49-3d6cc71ac1a9/repository/download/ConclaveCore_SdkBuild/3619:id/docs.zip!/kds-configuration.html

ac101m replied:
Yeah, intellij keeps removing the indentation here. Not sure why. It's been fixed.

shamsasari reviewed (changes requested):
Just blocking this

ac101m commented:
> I like the detailed and step-by-step instructions. Most of my comments are to match the content according to our style guide.

I've actually not read the style guide, I should probably do that! 
I addressed most comments as requested, but a few were done slightly differently. PTAL.

JohnJoser3 reviewed (approved):
Changes look good.

shamsasari commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +158,94 @@ kds {
+    For the R3 KDS, the following constraint should be used: `S:B4CDF6F4FA5B484FCA82292CE340FF305AA294F19382178BEA759E30E7DCFE2D PROD:1 SEC:STALE`.
+
+### Defining a KDS key spec
+To make use of KDS keys with Mail, the client must construct an appropriately configured [`PostOffice`](api/-conclave/com.r3.conclave.mail/-post-office/index.html) object with which to encrypt Mail.
```
Why this duplication of the phrase "To make use of KDS keys with Mail"?

ac101m replied:
Yeah, it's a bit meh. Removed.

shamsasari commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +158,94 @@ kds {
+    For the R3 KDS, the following constraint should be used: `S:B4CDF6F4FA5B484FCA82292CE340FF305AA294F19382178BEA759E30E7DCFE2D PROD:1 SEC:STALE`.
+
+### Defining a KDS key spec
+To make use of KDS keys with Mail, the client must construct an appropriately configured [`PostOffice`](api/-conclave/com.r3.conclave.mail/-post-office/index.html) object with which to encrypt Mail.
```
Again, reference the correct post office builder method

shamsasari commented on docs/docs/kds-configuration.md:
````diff
@@ -207,32 +158,94 @@ kds {
+```
+
+- `keyName` - The name of the key. This name is used during key derivation and if changed, will result in a different key being generated. This can be used to separate keys by purpose.
+- `masterKeyType` - The type of master key to use for the derivation (see [`MasterKeyType`](api/-conclave/com.r3.conclave.common.kds/-master-key-type/index.html)).
````
Link to the master key list, whatever that ends up being.

shamsasari commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +158,94 @@ kds {
+
+...
+var kdsURL = new URL("<kds-url>");
+var kdsEnclaveConstraint = new EnclaveConstraints("<kds-enclave-constraints>");
```
This is not valid code. The class is `EnclaveConstraint` and there's no string c'tor for it.

ac101m replied:
Yup, was missing the .parse, fixed!

shamsasari reviewed (changes requested):
Looking very good.

ac101m commented on docs/docs/kds-configuration.md:
```diff
@@ -1,47 +1,39 @@
-    start. Enclaves built in non-release modes will start but will use
-    a key derived from the signer of the enclave, bound to the CPU of the
-    system. This allows for development without a connection to the KDS.
+    If the KDS URL is not provided and the enclave is a simulation or debug mode enclave, a warning will be printed and the enclave will default to signer derived keys.
```
I strongly disagree. We do a lot of this in our documentation and although _technically_ correct it is the opposite of helpful. In this case, "signer" is a generic term, but "CPU key" and "MRSIGNER" are SGX specific terms which require more knowledge to properly contextualise. A new user with a cursory understanding of SGX (which I ostensibly still am) will just be confused by this.

ac101m commented on docs/docs/kds-configuration.md:
```diff
@@ -1,47 +1,39 @@
-consists of parameters that specify the URL to use to connect to the KDS, and a
-duration after which an unresponsive connection to the KDS will timeout.
+## Configuring the KDS URL
+For an enclave to obtain keys from a KDS, the URL of a KDS instance must be provided when the enclave is started.
```
The ordering here is actually deliberate, as the KDS URL has to be specified for either storage or mail keys to be used. I've added some wording to this effect.

ac101m commented on docs/docs/kds-configuration.md:
```diff
@@ -54,147 +46,106 @@ enclaveHost.start(null, null, null, kdsConfiguration) {}
-the key specification is not tampered with.
+## KDS keys for persistent storage
+To use KDS derived keys for persistent storage, you need to specify how to derive the storage key.
+Key specifications are configured on a per-enclave basis and are included in the enclave build configuration.
```
I've added a link to the bottom of the section.
I should say though that I think the term "key specification" is general enough to be understandable on its own, especially when framed by the preceding sentence. I also wrote this section with the understanding that this doc would be first port of call for anyone looking to use the KDS, so I think the breakdown covers it pretty well.

ac101m commented on docs/docs/kds-configuration.md:
```diff
@@ -54,147 +46,106 @@ enclaveHost.start(null, null, null, kdsConfiguration) {}
+- `persististenceKeySpec` - Subsection containing specifications for how encrypted storage keys should be derived.
+
+#### Persistence key specification section (`persistenceKeySpec`)
+- `masterKeyType` - Which KDS master key should we use when deriving a key, supported values are `development`, `cluster` and `azure_hsm`.
```
Done, also added links to the API docs, which I think also do a pretty good job of explaining the key types.

ac101m commented on docs/docs/kds-configuration.md:
```diff
@@ -54,147 +46,106 @@ enclaveHost.start(null, null, null, kdsConfiguration) {}
+    For the R3 KDS, the following constraint should be used: `S:B4CDF6F4FA5B484FCA82292CE340FF305AA294F19382178BEA759E30E7DCFE2D PROD:1 SEC:STALE`.
+
+### Choosing a value for the `masterKeyType`
+The KDS contains several internal master keys which may be used when deriving keys for user enclaves.
```
I've removed this sentence and done a little rewording.

ac101m commented on docs/docs/kds-configuration.md:
```diff
@@ -54,147 +46,106 @@ enclaveHost.start(null, null, null, kdsConfiguration) {}
+The `useOwnCodeHash` and `useOwnCodeSignerAndProductID` properties of the `policyConstraint` section allow you to add the code hash or code signer and product ID of the current enclave in the `constraint` property.
+These parameters cannot be added to the `constraint` field directly as their values are not known at build time.
+
+- The `useOwnCodeHash` property, when set to `true`, will restrict access to the key so that it is only released to enclaves with a code hash that matches the code hash of the current enclave.
```
Made it more obvious that the enclave properties are appended to the constraint.

ac101m commented on docs/docs/kds-configuration.md:
```diff
@@ -54,147 +46,106 @@ enclaveHost.start(null, null, null, kdsConfiguration) {}
+The master key is an internal secret of the KDS instance that is used during key derivation to prevent other KDS instances from deriving the same keys.
+There are several supported key types, each with different trust models:
+
+- `development` - A stable master key which is not suitable for production workloads, but can be used to test KDS integration. Release mode enclaves cannot use this key type.
```
I think at the moment the kds-detail doc is a little redundant in this respect. This document is intended to be mostly practical/instructional, with a little background information.
I think the best course is to link to the kds-detail doc, and flesh out the description of the key types there to include more background information about how they work and what the differences are.

ac101m commented on docs/docs/kds-configuration.md:
```diff
@@ -54,147 +46,106 @@ enclaveHost.start(null, null, null, kdsConfiguration) {}
-uses the enclave to continue running whilst giving the system administrator time
-to deploy the patch to the system, bringing it back to the `SECURE` state.
+This is a restrictive policy constraint with significant operational implications.
+Because of the code hash requirement, persisted data will be inaccessible to subsequent versions of the enclave.
```
The word "upgraded" is very broad. When we say "upgraded", what we really mean is something much more specific, something like "replacing the enclave jar in-place and maintaining access to previously persisted data". Stating this again is redundant.

ac101m commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +158,94 @@ kds {
+    In this example, security may be enhanced by having an external auditor build and sign the enclave on the behalf of the developer.
+
+## KDS keys for Conclave Mail
+The KDS is also capable of providing keys for use with [Conclave Mail](mail.md) through use of a specially constructed [`PostOffice`](api/-conclave/com.r3.conclave.mail/-post-office/index.html).
```
The API docs for usingKDS are linked later in the doc at the point when the reader is instructed to interact with the method. Also, it's irritating to click on a method in a code fragment and be taken to somewhere other than the class/method indicated by the text in the link.
I've adjusted the link later in the doc to make it a bit better.

shamsasari commented on docs/docs/kds-configuration.md:
```diff
@@ -10,30 +10,30 @@ In addition to CPU based keys for storage, Conclave makes use of random session
-    The URL for the R3 public KDS is `https://kds.dev.conclave.cloud`.
-    This URL may be changed in a later deployment of the KDS.
+    R3 runs a development KDS at the following URL: `https://kds.dev.conclave.cloud`.
+    A production ready deployment is pending.
```
will be available soon.

ac101m replied:
Done

shamsasari commented on docs/docs/kds-configuration.md:
```diff
@@ -207,32 +164,94 @@ kds {
+The KDS is also capable of providing keys for use with [Conclave Mail](mail.md) through use of a specially constructed [`PostOffice`](api/-conclave%20-core/com.r3.conclave.mail/-post-office/index.html).
+
+When using a KDS post office, the post office requests a public key from the KDS for use in a Conclave [`PostOffice`](api/-conclave%20-core/com.r3.conclave.mail/-post-office/index.html) and specifies which enclaves should be permitted access to the corresponding private key.
+Upon receiving Mail, the enclave requests the private key from the same KDS and uses it to decrypt the Mail object from the client.
```
Mail message

ac101m replied:
Whoops, thought I got all of these.

shamsasari reviewed (approved):
Awesome
